plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1359 @@
+        n: Union[Callable, u.Quantity] = None,
+        time_input: u.Quantity = None,
+        time_start: u.Quantity = None,
+        time_max: u.Quantity = None,
+        max_steps: Union[int, np.integer] = 10000,
+        tol: Union[int, float] = 1e-15,
+        dt: u.Quantity = None,
+        dt_max: u.Quantity = np.inf * u.s,
+        dt_min: u.Quantity = 0 * u.s,
+        adapt_dt: bool = None,
+        safety_factor: Union[int, float] = 1,
+        verbose: bool = False,
+    ):
+
+        try:
+
+            self.time_input = time_input
+            self.time_start = time_start
+            self.time_max = time_max
+            self.T_e_input = T_e
+            self.n_input = n
+            self.max_steps = max_steps
+            self.dt_input = dt
+
+            if self.dt_input is None:
+                self._dt = self.time_max / max_steps
+            else:
+                self._dt = self.dt_input
+
+            self.dt_min = dt_min
+            self.dt_max = dt_max
+            self.adapt_dt = adapt_dt
+            self.safety_factor = safety_factor
+            self.verbose = verbose
+
+            T_e_init = self.electron_temperature(self.time_start)
+            n_init = self.hydrogen_number_density(self.time_start)
+
+            self.initial = IonizationStates(
+                inputs=inputs, abundances=abundances, T_e=T_e_init, n=n_init, tol=tol,
+            )
+
+            self.tol = tol
+
+            # TODO: Update IonizationStates in PlasmaPy to have elements attribute
+
+            self.elements = list(self.initial.ionic_fractions.keys())
+
+            if "H" not in self.elements:
+                raise NEIError("Must have H in elements")
+
+            self.abundances = self.initial.abundances
+
+            self._EigenDataDict = {
+                element: EigenData(element) for element in self.elements
+            }
+
+            if self.T_e_input is not None and not isinstance(inputs, dict):
+                for element in self.initial.ionic_fractions.keys():
+                    self.initial.ionic_fractions[element] = self.EigenDataDict[
+                        element
+                    ].equilibrium_state(T_e_init.value)
+
+            self._temperature_grid = self._EigenDataDict[
+                self.elements[0]
+            ].temperature_grid
+
+            self._get_temperature_index = self._EigenDataDict[
+                self.elements[0]
+            ]._get_temperature_index
+
+        except Exception:
+            raise NEIError(
+                f"Unable to create NEI instance for:\n"
+                f"     inputs = {inputs}\n"
+                f" abundances = {abundances}\n"
+                f"        T_e = {T_e}\n"
+                f"          n = {n}\n"
+                f" time_input = {time_input}\n"
+                f" time_start = {time_start}\n"
+                f"   time_max = {time_max}\n"
+                f"  max_steps = {max_steps}\n"
+            )
+
```

Comment:
This is a large file and :bread: won't make itself. I'll be back to it in a bit!